### ISSUE TITLE
viz testing B1phi1 data

### ISSUE THREAD

**Issue by subwaystation** [vgteam/odgi#71]
Input was [run1.B1phi1.i1.seqwish.gfa.tar.gz](https://github.com/vgteam/odgi/files/4103860/run1.B1phi1.i1.seqwish.gfa.tar.gz).
Today's master of odgi. 
Today's master of component_segmentation.
Today's master of Schematize.

Then I executed 
```bash
#!/bin/bash

ODGI=~/software/odgi/git/master/bin/odgi
GFA=run1.B1phi1.i1.seqwish.gfa
OG=${GFA%.gfa}.og
SOG=${GFA%.gfa}.sorted.og


## Build the sparse matrix form of the gfa graph
echo "### odgi build"
BLDPREF=${0%.sh}_01_build
/usr/bin/time -v -o ${BLDPREF}.time \
ionice -c2 -n7 \
$ODGI build \
--progress \
--gfa=$GFA \
--out=$OG \
> ${BLDPREF}.log 2>&1

## Sort paths by 1D sorting
echo "### odgi sort"
SRTPREF=${0%.sh}_02_sort
/usr/bin/time -v -o ${SRTPREF}.time \
ionice -c2 -n7 \
$ODGI sort \
--pipeline="bSnSnS" \
--sgd-use-paths \
--paths-max \
--progress \
--idx=$OG \
--out=$SOG \
> ${SRTPREF}.log 2>&1

echo "### odgi bin"
for w in 100 1000 10000 100000; do
	BIN=${GFA%.gfa}.w${w}.json
	BINPREF=${0%.sh}_04_bin_w${w}
	/usr/bin/time -v -o ${BINPREF}.time \
	ionice -c2 -n7 \
	$ODGI bin \
	--json \
	--idx=$SOG \
	--bin-width=${w} \
	1> $BIN \
	2> ${BINPREF}.log &
done
```
on our VM pantograph2. If you want access and fiddle around their directly, just tell me @ekg 
Next step: graph segmentation into components:
`python component_segmentation/matrixcomponent/segmentation.py -j ~/sh/11_odgi/run1.B1phi1.i1.seqwish.w100.json -o ~/sh/11_odgi/`
and plugged the resulting file into our Schematize React application.
Screenshot with first bin:
![Screenshot from 2020-01-23 16-35-07](https://user-images.githubusercontent.com/10184961/73000041-94777400-3e00-11ea-8256-ad551c2a9d52.png)
Screenshot without first bin:
![Screenshot from 2020-01-23 16-39-51 - without first bin](https://user-images.githubusercontent.com/10184961/72999975-76aa0f00-3e00-11ea-846e-0436da124917.png)

If you need anything else, I'd be happy to help.

**Comment by josiahseaman** [comment]
To fix this, in component segmentation I added:
```
    if 0 in [upstream, downstream]:
         continue  # ignore links to the telomere.
```
After eliminating connections to the first bin, the whole file goes from 98MB to 4MB.  There's 60 bins, with very little variation between them.

**Comment by ekg** [comment]
That's really weird. I guess something is wrong with binning.

**Comment by ekg** [comment]
@subwaystation There is some basic confusion about the format that we should clear up. The bins are numbered from 1:N. The links in each path record refer to '0' to indicate the start and end of the path. I'm not sure if this is by design or accident.

I can change this format, but it does seem to be sufficient at present. You were right then to remove links to or from 0.

Is there any reason to leave these links? It seems that we know the path start and end based on the list of bins it goes through. I think it will just cause further confusion.

**Comment by ekg** [comment]
Actually, it might be necessary to retain it. I don't see any other way to determine where the paths start or end. The path bins just list which bins the paths touch.

**Comment by subwaystation** [comment]
Quoting @ekg `The link is from 0 to the bin where the path starts or from the path end bin to 0 for every path. It's different for everyone.`

That means our current segmentation algorithm only cares about structural variation in general, but not how our components look on the inside. We have to make that clear in our viz. Else people could get the impression, every path is present in every bin.

**Comment by ekg** [comment]
This is correct. And the bin width determines the resolution at which we can see structural variants.

**Comment by josiahseaman** [comment]
Thanks for the clarification on the format @ekg.  I guess the bug was actually in our segmenter all along.  I've corrected it now so I think you can close this issue.  Is there any indication of the end of a path?  Is this unrelated to us seeing a lot of connections between the ends of Arabidopsis chromosomes?

**Comment by ekg** [comment]
Chromosome ends will be likely to have SVs and homologies to each other.
Filtering this out can be done by changing the way the graph is built. I'd
say, check, though to be certain it's not the same problem.

**Comment by josiahseaman** [comment]
Does the end of a path loop back to Bin 0?

**Comment by subwaystation** [comment]
Also, the Arabidopsis data we are currently visualizing was not generated with @ekg's new fancy SGD in 1D sorting method, but the old one. Will have to do that.

**Comment by ekg** [comment]
0 is just a marker in the binned output to clarify when the startbor end of a path is in a certain bin.

**Comment by josiahseaman** [comment]
Thanks, I think we can close this issue now.

**Comment by subwaystation** [comment]
I have another question. If I got it correctly, all endings of all the paths are specified in this 0 bin. So only the paths' endings are specified, but not there beginning?
I am asking because I am currently implementing task 1 of graph-genome/Schematize#9, the adjacent connectors. @josiahseaman implemented a way to keep track of all paths in each component https://github.com/graph-genome/component_segmentation/commit/aab2e7534b0e1b44c1d9e60b3cdbb48fe5c4ebe0. This blows up the JSON, of course. 
Having both the beginnings and endings already available, we could circumvent that. 
But as far as I got it, that is not the case @ekg ?

**Comment by josiahseaman** [comment]
Sorry I think there's a misunderstanding.  Paths always start with 0 bin and end with 0 bin, so we do have beginnings and ends marked in the data.  Occupancy was just a way for me to speed up checking if a Path was using a particular Bin.  This is also covered by the matrix.  So I don't think there's anything that we need added on the odgi side.

**Comment by subwaystation** [comment]
Thanks, I got it now. The first link always describes the starting bin `[0, startBin]` and the last link in the array the ending bin `[endBin, 0]`.

One more interesting thing I ran into e.g. https://github.com/graph-genome/component_segmentation/blob/098069025e205563be2f4cd4e7c9f05f1256bdd1/matrixcomponent/data/run1.B1phi1.i1.seqwish.w100.json#L2461:
`"links":[[0,62],[58,56],[50,48],[48,0]]}`.
We start in a larger bin than we end. End all links basically go from left to right.
So the whole sequence of that phage must be read backward. So far I assumed the start bin would always be the lowest one. Good to know!
Is this a result of the sorting @ekg ?

**Comment by ekg** [comment]
I think it's that some of the paths are inverted relative to the default
orientation of the graph.

There should be a field giving the inversion state. It's in the tabular bin
output.

**Comment by subwaystation** [comment]
Some more sorting results for comparison @ekg @tpook92.
Below the simple sort, which is the exact same sort we did in the first round for the A. thaliana data.

```bash
## Sort paths by their average contained node id
echo "### odgi sort"
SRTPREF=${0%.sh}_02_sort
/usr/bin/time -v -o ${SRTPREF}.time \
ionice -c2 -n7 \
$ODGI sort \
--paths-avg \
--progress \
--idx=$OG \
--out=$SOG \
> ${SRTPREF}.log 2>&1
```

![run1 B1phi1 i1 seqwish sorted og R simple](https://user-images.githubusercontent.com/10184961/74517582-a2f21080-4f12-11ea-80e4-1ada60bef0cc.png)

**Comment by subwaystation** [comment]
Now with the new 1D sorting @ekg implemented.

```bash
## Sort paths by 1D sorting
echo "### odgi sort"
SRTPREF=${0%.sh}_02_sort
/usr/bin/time -v -o ${SRTPREF}.time \
ionice -c2 -n7 \
$ODGI sort \
--pipeline="bSnSnS" \
--sgd-use-paths \
--paths-max \
--progress \
--idx=$OG \
--out=$SOG \
> ${SRTPREF}.log 2>&1
```

![run1 B1phi1 i1 seqwish sorted ogR K bSnSnS](https://user-images.githubusercontent.com/10184961/74517953-5c50e600-4f13-11ea-9029-cb9581903e03.png)

The difference is huge. The result with 1D sorting just amazing.
But the A. thaliana data was much more complex, so we may have to adjust here. Curious about your results @tpook92.

**Comment by ekg** [comment]
@subwaystation to be clear, all the sorts are 1D.

This approach in particular is based on a stochastic gradient descent layout algorithm for graphs described here: https://www.repository.cam.ac.uk/handle/1810/294516. "Graph Drawing by Stochastic Gradient Descent"

It's been rewritten in odgi to be fully parallel, using atomic primitives for lock-free coordination. I've also moved it to 1D. An interesting project would be to port the parallel implementation back to 2D.

The major problem is that the data structures are very expensive. We end up storing a lot of node pairs and some distance and weight information for each. We have to also hold onto an atomic primitive for each pairing. Getting the sampling of node pairs to be more sparse, but in the right way, could really help the memory and runtime performance of the method.

**Comment by ekg** [comment]
Amending the previous comment with some more algorithm details...

To compute the SGD layout algorithm, we would ideally have all pairs of nodes and the actual graph-based distance between them. This could be obtained with a breadth first search from every node to the rest of the graph. Obviously, the runtime cost to determine this is O(N^2), and so is the memory required to store the distances. For each term (pair of nodes), we then have an actual distance in the graph and a layout distance in the 1D (or 2D) layout. The SGD updates pick a random term and shift the node distance in the layout space to better match the actual distance between them. The rate at which this happens is controlled by a learning rate parameter that decays with (approximate) iterations over the whole data set.

Here's the key figure describing this:

![Screenshot from 2020-02-14 10-51-50](https://user-images.githubusercontent.com/145425/74520535-10ed0680-4f18-11ea-9e90-2ac77ec75d56.png)

Using this layout algorithm for graphs is brilliant, and it's amazing that no one else had explored it previously. The layout quality is excellent. To my knowledge, it's one of the few major advances in graph drawing in the past decade.

**Comment by tpook92** [comment]
This is the old AB dataset without sorting (Regions in black indicate regions with present variation but not assigned to haplotype blocks): 

![sebastianOLD_nobreak Athaliana all 50000 w1000 json](https://user-images.githubusercontent.com/38560663/74521081-1bf46680-4f19-11ea-9f80-bdd8ec35f3c5.png)

And this is the new dataset:
![sebastianNEW_nobreak Athaliana all 50000 w1000 json](https://user-images.githubusercontent.com/38560663/74521182-49411480-4f19-11ea-8e33-54dc7f2b77fb.png)

The larger chromosome chucks are a bit split BUT the regions not-assign to any genome is going down massively. 

In this is how it looks after sorting:
Old:
![sebastianOLD Athaliana all 50000 w1000 json](https://user-images.githubusercontent.com/38560663/74528038-7f859080-4f27-11ea-98aa-4580c64f9028.png)

New:
![sebastianNEW Athaliana all 50000 w1000 json](https://user-images.githubusercontent.com/38560663/74528049-86ac9e80-4f27-11ea-9538-1cfff96f8e78.png)

Overall i would say that the new sorting is leading to better results - i am not 100% happy with my own sorting but thats a different story.

**Comment by ekg** [comment]
To get chromosomes to not sort together, it'd be enough to build the graph one chromosome at a time, sort each, and then concatenate the graphs. In odgi graph combination isn't yet well-supported from the CLI, but it'd be trivial to set up.

**Comment by subwaystation** [comment]
Would that still reflect the biological truth? Do we leave out chromosome cross talking (e.g. CNV on another chromosome)? Or would that be a desired effect?

**Comment by josiahseaman** [comment]
@ekg I second Simon's question.

@tpook92 Thank you for your hard work so far.  Are there parameters which you can change to get the haploblocks to more aggressively bridge the white space to include larger sections?  So far, it looks like the haploblocks are too small to be meaningful.  Is there a setting where you get one giant haploblock for the whole chromosome?  Can you find a useful midpoint between the two?

**Comment by tpook92** [comment]
@josiahseaman i put some more weight on the detection on longer blocks and are know allowing for overlapping blocks (and removed a typo from my side...). The average blocks length for the 1000w dataset is now 600 bins (median ~350). My blocks are covering ~70% of all present variation.

![hb_update](https://user-images.githubusercontent.com/38560663/74675601-2d0cd400-51b4-11ea-94a3-da99a5f94751.jpg)

From visual inspection i would say that the parts not covered by haplotype blocks are mostly unique variants. I dont thinks its possible to detect much longer blocks in this dataset - lines are just too different from each other. Results will probably massively improve with more lines - we are supposed to get like 30 arabidopbis lines from Christian, right?

Sorting is still not ideal - in particular i am planning to put more weight on longer haplotype blocks instead of mostly one anchor region.

**Comment by subwaystation** [comment]
So that data you are currently working with is the 12 genomes from Christian?
